src/main/java/org/sonar/plugins/csharp/SarifParser.java

```diff
@@ -44,33 +48,59 @@ public void parse(File file) {
     }
 
     JsonParser parser = new JsonParser();
-    JsonElement issues = parser.parse(contents).getAsJsonObject().get("issues");
-    if (issues != null) {
-      for (JsonElement issueElement : issues.getAsJsonArray()) {
-        JsonObject issue = issueElement.getAsJsonObject();
+    JsonObject root = parser.parse(contents).getAsJsonObject();
+    if (root.has("runLogs")) {
+      JsonElement runLogs = parser.parse(contents).getAsJsonObject().get("runLogs");
+      for (JsonElement runLogElement: runLogs.getAsJsonArray()) {
+        JsonObject runLog = runLogElement.getAsJsonObject();
+        JsonArray results = runLog.getAsJsonArray("results");
+        if (results != null) {
+          handleIssues(results, false, true);
+        }
+      }
+    } else if (root.has("issues")) {
+      JsonElement issues = parser.parse(contents).getAsJsonObject().get("issues");
+      handleIssues(issues.getAsJsonArray(), true, false);
+    }
+  }
+
+  private void handleIssues(JsonArray issues, boolean linesOffByOne, boolean convertUri) {
+    for (JsonElement issueElement : issues) {
+      JsonObject issue = issueElement.getAsJsonObject();
 
-        String ruleId = issue.get("ruleId").getAsString();
+      String ruleId = issue.get("ruleId").getAsString();
 
-        String message = issue.get(issue.has("shortMessage") ? "shortMessage" : "fullMessage").getAsString();
-        boolean hasLocation = false;
-        for (JsonElement locationElement : issue.get("locations").getAsJsonArray()) {
-          JsonObject location = locationElement.getAsJsonObject();
-          if (location.has("analysisTarget")) {
-            for (JsonElement analysisTargetElement : location.get("analysisTarget").getAsJsonArray()) {
-              hasLocation = true;
-              JsonObject analysisTarget = analysisTargetElement.getAsJsonObject();
-              String uri = analysisTarget.get("uri").getAsString();
-              JsonObject region = analysisTarget.get("region").getAsJsonObject();
-              int startLine = region.get("startLine").getAsInt();
+      String message = issue.get(issue.has("shortMessage") ? "shortMessage" : "fullMessage").getAsString();
+      boolean hasLocation = false;
+      for (JsonElement locationElement : issue.get("locations").getAsJsonArray()) {
+        JsonObject location = locationElement.getAsJsonObject();
+        if (location.has("analysisTarget")) {
+          for (JsonElement analysisTargetElement : location.get("analysisTarget").getAsJsonArray()) {
+            hasLocation = true;
+            JsonObject analysisTarget = analysisTargetElement.getAsJsonObject();
 
-              callback.onIssue(ruleId, uri, message, startLine);
+            String uri = analysisTarget.get("uri").getAsString();
+            if (convertUri) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Refactor this code to not nest more than 4 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS134)